service/common/src/main/java/org/apache/polaris/service/events/listeners/InMemoryBufferPolarisPersistenceEventListener.java

```diff
@@ -0,0 +1,166 @@
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(InMemoryBufferPolarisPersistenceEventListener.class);
+  private static final String REQUEST_ID_KEY = "requestId";
+  private final MetaStoreManagerFactory metaStoreManagerFactory;
+  private final Clock clock;
+
+  private final ConcurrentHashMap<String, ConcurrentLinkedQueue<EventAndContext>> buffer =
+      new ConcurrentHashMap<>();
+  private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+  private final ConcurrentHashMap<Future<?>, Integer> futures = new ConcurrentHashMap<>();
+  private final Duration timeToFlush;
+  private final int maxBufferSize;
+
+  @Context ContainerRequestContext containerRequestContext;
+
+  private record EventAndContext(PolarisEvent polarisEvent, PolarisCallContext callContext) {}
+
+  @Inject
+  public InMemoryBufferPolarisPersistenceEventListener(
+      MetaStoreManagerFactory metaStoreManagerFactory,
+      Clock clock,
+      InMemoryBufferPersistenceListenerConfiguration eventListenerConfiguration) {
+    this.metaStoreManagerFactory = metaStoreManagerFactory;
+    this.clock = clock;
+    this.timeToFlush =
+        eventListenerConfiguration.bufferTime().orElse(Duration.of(30, ChronoUnit.SECONDS));
+    this.maxBufferSize = eventListenerConfiguration.maxBufferSize().orElse(5); // 5 events default
+  }
+
+  @PostConstruct
+  void start() {
+    futures.put(
+        executor.scheduleAtFixedRate(
+            this::runCleanup, 0, timeToFlush.toMillis(), TimeUnit.MILLISECONDS),
+        1);
+  }
+
+  void runCleanup() {
+    for (String realmId : buffer.keySet()) {
+      try {
+        checkAndFlushBufferIfNecessary(realmId);
+      } catch (Exception e) {
+        LOGGER.debug("Buffer checking task failed for realm ({}): {}", realmId, e);
+      }
+    }
+    // Clean up futures
+    try {
+      futures.keySet().removeIf(future -> future.isCancelled() || future.isDone());
+    } catch (Exception e) {
+      LOGGER.debug("Futures reaper task failed.");
+    }
+  }
+
+  @PreDestroy
+  void shutdown() {
+    futures.keySet().forEach(future -> future.cancel(false));
+    executor.shutdownNow();
+  }
+
+  @Override
+  String getRequestId() {
+    if (containerRequestContext != null && containerRequestContext.hasProperty(REQUEST_ID_KEY)) {
+      return (String) containerRequestContext.getProperty(REQUEST_ID_KEY);
+    }
+    return UUID.randomUUID().toString();
+  }
+
+  @Override
+  void addToBuffer(PolarisEvent polarisEvent, CallContext callCtx) {
+    String realmId = callCtx.getRealmContext().getRealmIdentifier();
+
+    buffer
+        .computeIfAbsent(realmId, k -> new ConcurrentLinkedQueue<>())
+        .add(new EventAndContext(polarisEvent, callCtx.getPolarisCallContext().copy()));
+    futures.put(executor.submit(() -> checkAndFlushBufferIfNecessary(realmId)), 1);
+  }
+
+  @VisibleForTesting
+  public void checkAndFlushBufferIfNecessary(String realmId) {
```

Comment:
This will run for each call to `addToBuffer()` - that's quite a lot of immediately scheduled futures.
I still suspect a lot of elements piling up in the queue(s), where processing of those is being stalled.
The implementation doesn't seem to be resilient against such scenarios, like millions of "rate limited" events to bring down the service.